jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

```diff
@@ -1811,6 +1811,11 @@ public boolean isUserInRole(String role)
      */
     public void setMetaData(org.eclipse.jetty.http.MetaData.Request request)
     {
+        if (_metaData == null && _input != null && _channel != null)
+        {
+            _input.recycle();
+            _channel.getResponse().getHttpOutput().reopen();
```

Comment:
The fact that output can be recycled early but needs to be reopened on the next cycle while the input cannot be recycled early and must be done on the next cycle is confusing as both sides don't share the recycle pattern equally anymore. I find that really confusing.

I do not critique the design of what's happening under the hood, rather the way this is exposed. I'd prefer if both input and output had a `reopen` method that would be called from the same spot and have their `recycle` method also called from the same spot.

Reply:
kind of agree, but I don't want to go adding state to HttpInput, as it is fragile enough.  How about you do this when we port to 10 in the new HttpInput?

Reply:
This will make porting future fixes more complicated, but maybe not that much more than now. Let's go with your proposed plan for now, we'll reevaluate if aligning 9.4 and 10 in this matter is worth it the next time we have to work on HttpInput.